src/raw/mod.rs

```diff
@@ -0,0 +1,1849 @@
+    data: Box<[u8]>,
+}
+
+impl DocBuf {
+    /// Create a new `DocBuf` from the provided `Vec`.
+    ///
+    /// The data is checked for a declared length equal to the length of the Vec,
+    /// and a trailing NUL byte.  Other validation is deferred to access time.
+    ///
+    /// ```
+    /// # use bson::raw::{DocBuf, RawError};
+    /// let docbuf: DocBuf = DocBuf::new(b"\x05\0\0\0\0".to_vec())?;
+    /// # Ok::<(), RawError>(())
+    /// ```
+    pub fn new(data: Vec<u8>) -> RawResult<DocBuf> {
+        if data.len() < 5 {
+            return Err(RawError::MalformedValue("document too short".into()));
+        }
+        let length = i32_from_slice(&data[..4]);
+        if data.len() as i32 != length {
+            return Err(RawError::MalformedValue("document length incorrect".into()));
+        }
+        if data[data.len() - 1] != 0 {
+            return Err(RawError::MalformedValue(
+                "document not null-terminated".into(),
+            ));
+        }
+        Ok(unsafe { DocBuf::new_unchecked(data) })
+    }
+
+    /// Create a DocBuf from a [bson::Document].
+    ///
+    /// ```
+    /// # use bson::raw::{DocBuf, RawError};
+    /// use bson::{doc, oid};
+    /// let document = doc! {
+    ///     "_id": oid::ObjectId::new(),
+    ///     "name": "Herman Melville",
+    ///     "title": "Moby-Dick",
+    /// };
+    /// let docbuf: DocBuf = DocBuf::from_document(&document);
+    /// # Ok::<(), RawError>(())
+    /// ```
+    pub fn from_document(doc: &crate::Document) -> DocBuf {
+        let mut data = Vec::new();
+        doc.to_writer(&mut data).unwrap();
+        unsafe { DocBuf::new_unchecked(data) }
+    }
+
+    /// Create a DocBuf from an owned Vec<u8> without performing any checks on the provided data.
+    ///
+    /// ```
+    /// # use bson::raw::{DocBuf, RawError};
+    /// let docbuf: DocBuf = unsafe {
+    ///     DocBuf::new_unchecked(b"\x05\0\0\0\0".to_vec())
+    /// };
+    /// # Ok::<(), RawError>(())
+    /// ```
+    ///
+    /// # Safety
+    ///
+    /// The provided bytes must have a valid length marker, and be NUL terminated.
+    pub unsafe fn new_unchecked(data: Vec<u8>) -> DocBuf {
+        DocBuf {
+            data: data.into_boxed_slice(),
+        }
+    }
+
+    /// Return a [`&Doc`](Doc) borrowing from the data contained in self.
+    ///
+    /// # Deprecation
+    ///
+    /// DocRef is now a deprecated type alias for [Doc].  DocBuf can
+    /// dereference to &Doc directly, or be converted using [AsRef::as_ref],
+    /// so this function is unnecessary.
+    ///
+    /// ```
+    /// # use bson::raw::{DocBuf, DocRef, RawError};
+    /// let docbuf = DocBuf::new(b"\x13\x00\x00\x00\x02hi\x00\x06\x00\x00\x00y'all\x00\x00".to_vec())?;
+    /// let docref: DocRef = docbuf.as_docref();
+    /// # Ok::<(), RawError>(())
+    /// ```
+    #[deprecated(since = "0.2.0", note = "use docbuf.as_ref() instead")]
+    pub fn as_docref(&self) -> &Doc {
+        self.as_ref()
+    }
```

Comment:
I recommend yanking out all mentions of DocRef. It doesn't need to exist anymore.

Reply:
Yep, that makes sense. I removed this method in my follow-up commit.